### DISCUSSION TITLE
Determining Warehouse Distribution Center Location

### DISCUSSION THREAD

**Discussion by ELITSZORD** [teamalgoritma/community#871] [Q&A] [answered]
Dear Algoritma Team, Mba Ranny, @IrfanChairurrachman 

Jika ingin menentukan lokasi strategis sebuah gudang, kebanyakan yang saya cari menggunakan algoritma K-means dan Weighted K-Means (apakah ada yang bisa menjelaskan Weighted K-Means, bingung mengenai weightedya ini)?

Kemudain, apakah ada rekomendasi algoritma lain yang bisa untuk case dalam menentukan pusat distribusi gudang dan berikut tutorialnya? 

Terimakasih.

**Comment by ranydwi** [comment, marked as answer]
Dear Pak @ELITSZORD , terima kasih untuk pertanyaannya dan kesediannya untuk menunggu.

Algoritma weighted K-Means ini setahu saya adalah K-Means juga namun perbedaan yang digunakan hanyalah pada perhitungan jaraknya. Biasanya untuk K-Means, perhitungan jarak yang dipakai adalah eucledian dan manhattan. Namun, untuk weighted, di dalam perhitungan jarak yang dipakai diberikan suatu bobot/weight tertentu Pak (biasanya ketika menghitung untuk titik dimana titik tersebut lebih diperhitungkan). Mengapa demikian?

> Karena, masing-masing data point bisa diberi weight (bobot) yang berbeda, sehingga data point yang dianggap lebih penting akan lebih mempengaruhi posisi centroid.

Sebagai contoh: Misalnya, Bapak sedang melakukan clustering pada data pelanggan untuk menentukan warehouse distribution. Data point bisa berupa atribut pembelian pelanggan. Jika berat diberikan lebih tinggi pada lokasi dimana pelanggan lebih sering membeli disuatu lokasi tertentu, maka centroid akan lebih condong ke arah lokasi pelanggan tersebut, sehingga cluster yang dihasilkan diharapkan bisa lebih baik untuk warhouse distribution.

Penjelasan di artikel ini saya rasa sangat mudah dipahami pak: https://medium.com/@dey.mallika/unsupervised-learning-with-weighted-k-means-3828b708d75d

Selain itu, algoritma yang mungkin bisa dicoba adalah Fuzzy C-Means Pak Faisal. Weighted ini sering di analogikan dengan fuzzy karena setiap penempatan centroid K tidak fixed melainkan terdapat probability untuk setiap penempatan centroid ke -K. Sehingga nantinya masih ada peluang suatu lokasi dijadikan sebagai warehouse center atau tidak. 

Sebagai contoh bisa dilihat melalui artikel berikut: https://rpubs.com/inayatus/fuzzy-c-means

Semoga membantu Pak!

**Reply by ELITSZORD** [reply]
Masya Allah terimakasih jawabannya mba Ranny sangat istimewa and crystal clear. Kemudian ada beberapa yang mau saya tanayakan:

1. Ketika menggunakan method `fit(X, y=None, sample_weight=None)` Bagaiaman kita memasukkan beberapa feature/series ke dalam parameter sample_weight dimana misal kita mempertimbangkan biaya distribusi, kelas jalan, dll? Tp ini numeric semua ya gk boleh factor atau kategoric?

2. Jika data setnya mixed antara numeric dan kategorik itu gimana ya? (Kalau menurut saya kemungkinan R akan otomatis one-hot encoding dan python tidak)

3. Dimana centroidnya kan akan berpindah2 kalau tidak di set.seed atau random.seed, bagaimana kita mengevaluasi bahwa centroid itu sudah bagus sebagai pusat distrbiusi di setiap cluster?

**Reply by ranydwi** [reply]
Dear Pak @ELITSZORD ,

1. Untuk method `fit(X, sample_weight = Y)`, parameter `sample_weight`  hanya dapat berisikan satu variabel yang berisi nilai bobot untuk setiap observasi dalam dataset. Meskipun demikian, Bapak dapat menggunakan kombinasi dari beberapa variabel untuk menghasilkan nilai bobot tunggal. Seperti metode aggregasi (mean, sum, median) maupun dengan menggunakan bobot per variabel. Source: https://scikit-learn.org/stable/modules/generated/sklearn.cluster.KMeans.html
2. K-Means hanya bisa datasetnya numerik Pak, kalau mau bapak bisa mencoba method clustering lainnya seperti K-Medodis/PAM clustering. Kalau di R setahu saya ada Pak package untuk bisa menghitung weighted  K-Medoids bisa dibaca disini ([dokumentasi resmi](https://rdrr.io/cran/WeightedCluster/man/kmedoids.html )) maupun discussionnya [disini](https://datascience.stackexchange.com/questions/82658/k-medoid-clustering-with-point-weights)
3. Metode Clustering sendiri karena dia Unsupervised, maka sulit Pak untuk bisa melakukan evaluasi dari modelnya. Bapak bisa coba menghitung nilai WSS/BSS/TSS sebagai bagian dari pengecekan evaluasi

semoga membantu Pak!

**Reply by ELITSZORD** [reply]
Dear Mba Ranny @ranydwi ,    
1. Jadi pembobotan setiap observasi bisa dilakukan hanya dengan ke tiga agregasi itu saja ya..
2. Bener mba maksudnya jika di python untuk melakukan k-medoid harus di one-hot encoding dulu yaa untuk yang kategoric?
3. Ohh iya bener dengan WSS semakin rendah artinya tiap cluster memiliki karakteristik yang semakin mirip dan BSS/TSS mendekati 1 (namun tidak menjamin cluster itu sudah baik).
4. Jika case nya seperti mba ranny sebutkan
`Sebagai contoh: Misalnya, Bapak sedang melakukan clustering pada data pelanggan untuk menentukan warehouse distribution.` 

Bagaimana kita memplotkan kedalam suatu map ? apakah letak geographic suatau data observasi di masukkan saat dilakukan clustering atau ada cara lain mba yang benar?

**Reply by ranydwi** [reply]
Dear Pak @ELITSZORD thank you for your response and understanding to wait.
1. Untuk pertanyaan no 1 ini, Bapak bisa menggunakan metode agregasi yang lain, bahkan untuk metode normalisasi pun bisa digunakan. Hal ini bertujuan untuk mendapatkan 1 nilai kombinasi dari beberapa variabel yang tidak diakomodir oleh Weighted K-Means. Bapak bisa coba menggunakan MCDM method dengan AHP -> _Analytical Hierarchy Process_ contohnya (disini Bapak bisa sertakan bobot untuk masing-masing variabel)
2. Untuk memanfaatkan K-Medoids Clustering, tidak perlu yaa Pak melakukan one hot encoding variabel categoric di python
3. Betul sekali
4. Visualisasi Mapping bisa digunakan pada saat Bapak selesai melakukan clustering, sehingga nantinya output label yang bapak jadikan sebagai acuan. Atau apabila ada bobot tertentu, Bapak bisa masukan sebagai analisis EDA juga dengan mapping area yang dirasa cukup "diutamakan" baru hasilnya bisa dijustifikasi dengan output label hasil clustering ya Pak.

Regards,
Rany.
